peer/peer.go

```diff
@@ -0,0 +1,1942 @@
+	// too old.
+	if msg.ProtocolVersion < int32(wire.MultipleAddressVersion) {
+		// Send a reject message indicating the protocol version is
+		// obsolete and wait for the message to be sent before
+		// disconnecting.
+		reason := fmt.Sprintf("protocol version must be %d or greater",
+			wire.MultipleAddressVersion)
+		p.PushRejectMsg(msg.Command(), wire.RejectObsolete, reason,
+			nil, true)
+		p.Disconnect()
+		return
+	}
+
+	// Updating a bunch of stats.
+	p.StatsMtx.Lock()
+
+	// Limit to one version message per peer.
+	if p.versionKnown {
+		p.LogError("Only one version message per peer is allowed %s.",
+			p)
+		p.StatsMtx.Unlock()
+
+		// Send an reject message indicating the version message was
+		// incorrectly sent twice and wait for the message to be sent
+		// before disconnecting.
+		p.PushRejectMsg(msg.Command(), wire.RejectDuplicate,
+			"duplicate version message", nil, true)
+
+		p.Disconnect()
+		return
+	}
+
+	// Negotiate the protocol version.
+	p.protocolVersion = minUint32(p.protocolVersion, uint32(msg.ProtocolVersion))
+	p.versionKnown = true
+	log.Debugf("Negotiated protocol version %d for peer %s",
+		p.protocolVersion, p)
+	p.lastBlock = msg.LastBlock
+	p.startingHeight = msg.LastBlock
+
+	// Set the supported services for the peer to what the remote peer
+	// advertised.
+	p.services = msg.Services
+
+	// Set the remote peer's user agent.
+	p.userAgent = msg.UserAgent
+
+	// Set the peer's time offset.
+	p.timeOffset = msg.Timestamp.Unix() - time.Now().Unix()
+
+	// Set the peer's ID.
+	p.id = atomic.AddInt32(&nodeCount, 1)
+
+	p.StatsMtx.Unlock()
+
+	// Choose whether or not to relay transactions before a filter command
+	// is received.
+	p.SetDisableRelayTx(msg.DisableRelayTx)
+
+	// Inbound connections.
+	if p.inbound {
+		// Set up a NetAddress for the peer to be used with AddrManager.
+		// We only do this inbound because outbound set this up
+		// at connection time and no point recomputing.
+		na, err := newNetAddress(p.conn.RemoteAddr(), p.services)
+		if err != nil {
+			p.LogError("Can't get remote address: %v", err)
+			p.Disconnect()
+			return
+		}
+		p.na = na
+
+		// Send version.
+		err = p.pushVersionMsg()
+		if err != nil {
+			p.LogError("Can't send version message to %s: %v",
+				p, err)
+			p.Disconnect()
+			return
+		}
+	}
+
+	// Send verack.
+	p.QueueMessage(wire.NewMsgVerAck(), nil)
+
+	// Update the address manager and request known addresses from the
+	// remote peer for outbound connections.  This is skipped when running
+	// on the simulation test network since it is only intended to connect
+	// to specified peers and actively avoids advertising and connecting to
+	// discovered peers.
+	if !p.cfg.SimNet {
+		p.updateAddresses(msg)
+	}
+
+	// TODO: Relay alerts.
+}
+
+// AddVersionMsgListener adds a listener which is invoked when a peer receives
+// a version bitcoin message.
+func (p *Peer) AddVersionMsgListener(key string, listener func(p *Peer, msg *wire.MsgVersion)) {
```

Comment:
We either need to protect the callback registration and invocation by a mutex or clearly document that registration and invocation is not concurrent-safe.